### PR TITLE
Coding Tracker finally done!

### PR DESCRIPTION
**Pull request by AhmadJer99**
I struggled alot and got stuck alot of times during the making of this project, but alast its finally done!
Please if you see any issues even the smallest ones give feedback  its the art of failing!  i want to learn my mistakes.

### CODE REVIEW

**Review by TheCSharpAcademy — CHANGES_REQUESTED** [review]
@AhmadJer99  Thanks for submitting! 👌🏻At the moment I can add an end date before an end date, which violates one of our requirements: 
image

Also for better UX it would be good to let the users add the date as well, not just the time, or they'd be forced to input on the same day without fail.

**Comment by AhmadJer99** [comment]
Hey thanks for the review 🫡!
Shoot i totally missed fixing the issue "End Time before Start Time" in the console app even though i handled it in the .net maui version of the application, anyways i done fixed that now in the console app,

I also added the ability for the user to enter the date manually!

though i have 1 question about the date, should i allow multiple entries for the same date i mean multiple sessions for the same day and if i do allow it, should i have a unique constraint on the start time and  a session's date

meaning if a user tried to enter a new session on the same day and one already exists with the same start time it would return an error

**Comment by AhmadJer99** [comment]
Yeah and sorry the code is tangled alot in the main controller , i think i need to work on managing the code better and tidying it better than that.

File: CodingTracker.AhmadJer99/README.md
Comment (chrisjamiecarter) [inline CodingTracker.AhmadJer99/README.md:1]: 🟢 **README File**

💡 Well done on writing this, perfect introduction to your application.

File: CodingTracker.AhmadJer99/CodingTracker/Controllers/GenerateRandomData.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
+﻿using System.Configuration;
+using CodingTracker.Models;
+namespace CodingTracker.Controllers;
+
+internal static class GenerateRandomData
+{
+    public static int RandomRowsNumber { get; set; }
+    private readonly static string? dateFormat = ConfigurationManager.AppSettings.Get("CorrectDateFormat");
+    private readonly static string? timeFormat = ConfigurationManager.AppSettings.Get("CorrectTimeFormat");
+    private readonly static Random gen = new();
```
Comment (chrisjamiecarter) [inline CodingTracker.AhmadJer99/CodingTracker/Controllers/GenerateRandomData.cs:10]: 🟠 **Thread-Safe Random**

💡 From .NET 6 you can use `Random.Shared` to get a thread-safe instance of `Random`.

🔧 To get a random integer you just need to do: `int number = Random.Shared.Next();`

File: CodingTracker.AhmadJer99/CodingTracker/Controllers/OperationController.cs [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+        var codingSessions = DatabaseController.ReadAllRows(); // parse the rows into a list of coding sessions , each element is  a row in the table.
+
+        var columnNames = new List<string>() { "Id", "Session's Date", "Start Time", "End Time", "Duration(Hrs:Mins)" };
+        TableVisualisationEngine.ViewAsTable(codingSessions, TableAligntment.Center, columnNames);
+        // Pass the list of sessions into a table view engine , and show the results in a user friendly table.
+    }
+    internal static void FilteredView()
+    {
+        string filter = "";
```
Comment (chrisjamiecarter) [inline CodingTracker.AhmadJer99/CodingTracker/Controllers/OperationController.cs:165]: 🟠 **Multiple String Concatenations**

💡 Contentious topic, but general rule of thumb is:
- When concatenating three dynamic string values or less, use traditional string concatenation.
- When concatenating more than three dynamic string values, use `StringBuilder`.

🔧 `var builder = new StringBuilder(); builder.Append("..."); filter = builder.ToString();`

File: CodingTracker.AhmadJer99/CodingTracker/Properties/launchSettings.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+  "profiles": {
+    "CodingTracker": {
+      "commandName": "Project",
+      "workingDirectory": "C:\\Users\\AhmadJerjees\\source\\repos\\CodingTracker\\CodingTracker"
```
Comment (chrisjamiecarter) [inline CodingTracker.AhmadJer99/CodingTracker/Properties/launchSettings.json:5]: 🟠 **Hard-Coded Paths**

💡 How many machines in the world have this path? Only 1? ... 🤔 

🔧 Make sure you include this as a prerequisite step to change. Document it in the README.

File: CodingTracker.AhmadJer99/StopWatch/Program.cs
Comment (chrisjamiecarter) [inline CodingTracker.AhmadJer99/StopWatch/Program.cs:1]: 🟠 **Unused Code**

💡 Looks like this is a redundant project and should have been removed.

File: CodingTracker.AhmadJer99/StopWatch/StopWatch.csproj
Comment (chrisjamiecarter) [inline CodingTracker.AhmadJer99/StopWatch/StopWatch.csproj:1]: 🟠 **Unused Code**

💡 Looks like this is a redundant project and should have been removed.

File: CodingTracker.AhmadJer99/CodingTracker/Controllers/OperationController.cs [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+        var userTimeEntryChoice = AnsiConsole.Prompt(
+            new SelectionPrompt<StartTimeOptions>()
+            .Title("[yellow]Do you want to start the session manually by entering start and end date or you can use a stopwatch[/]")
+            .AddChoices(Enum.GetValues<StartTimeOptions>()));
+
+        switch (userTimeEntryChoice)
+        {
+            case StartTimeOptions.StopWatch:
+                Console.WriteLine("Unsupported Yet");
```
Comment (chrisjamiecarter) [inline CodingTracker.AhmadJer99/CodingTracker/Controllers/OperationController.cs:115]: 🟠 **Misleading Comments**

💡 This seems supported to me?

**Review by chrisjamiecarter — APPROVED** [review]
Hey @AhmadJer99 👋,

Fantastic work on your Coding Tracker project submission 🎉 !

I have performed a peer review and have added some praise/suggestions as comments, but also see some general feedback below.

🟢 **Solution Structure**
Well done on submitting your project with the correct structure. This has made it easy for me to pull down and review your code.

🟢 **Data Seeding**
Excellent job in seeding test data. This has allowed me to quickly test your application.

🟢 **Challenges Requirements**
I can:
- track my coding time via a stopwatch!
- filter my coding records per period and/or order them.
- see my total and average coding sessions per period.
- set a coding goal and see how far I am from reaching my goal.

🟠 **Style Guide Compliance**
Remember that when you have finished a project, give it a polish pass. Pay attention to any code conventions in place, in this instance refer to the C# Academy's [code-conventions](https://thecsharpacademy.com/article/58/code-conventions) page.
This is important as good code conventions/styling ensures consistency, makes your code easier for others to read and maintain, and improves your chances of getting a pull request reviewed and approved.
NOTE: Pay attention to spacing in your code. Separate your methods. Try not to make everything one big blob.

🟠 **User Experience**
Any reason these are PascalCase and not split withe spaces? i.e. `New Session / View All Sessions`

image

I don't know if this is just me or for everyone, but when I scroll your lists, in update for example, the title is spammed:

image

When updating a record, it would be good to know what I am updating. Why do I have to remember what was previously entered? That is what a database is for! To remember for me. Show the record being updated.

image

When choosing from a list, update for example, you have to select a record to 'cancel'. What if the user clicks update but meant to click something else. They are committed.

Sessions added via the stopwatch are missing the Date:

image

When you start a stopwatch session, end and save. Then go back into the stopwatch it restarts from the same time as when you last left. i.e. I add a 5 min stopwatch session, save and go back to main menu, go back to stopwatch session, and it restarts straight away from 5 mins.

***

I suggest revisiting to try understand and fix any issues 🔧 , but as these do not relate to the requirements they will not block my approval.

Feel free to reach out if you have any questions or if you'd like to collaborate further on any of these points 🆘 .

I will go ahead and mark as approved, keep up the excellent work on the next projects! 😊

Best regards,
@chrisjamiecarter 👍

**Comment by AhmadJer99** [comment]
Regarding why the user interface options where PascalCase and not split by a white space is becaused i used Enums  with ansi console selection prompt , i will not use it in my next projects  i will try to implement normal choice entry selection.

And no ,  you are not the only one experiencing this bug , it happens when you go windows mode with the console. If you have the console as full screen it wont spam as i have tried before. 
Idk why it keeps happening but i think the main reason is how AnsiConsole behaves.

I updated the update method to show which session you're currently updating when you are in the update menus.

I fixed all the menu so you can click back and not be committed to whatever you may click 

I fixed the issue where the stop watch saves sessions without dates

I understand that i need to use string builder since its more efficient and cleaner for the memory to work with a string builder instead of just concatenating , now i updated the filteredview method to use the string builder for the filter string .

The hard coded path totally caught me off guard,   I  didnt think about it being there tbh , thats on me

**Comment by AhmadJer99** [comment]
I fixed all unused code and the "unsupported" console writeline

**Comment by AhmadJer99** [comment]
Thanks for the amazing review Mr.Chris!

**Pull request closed without merging**